### DISCUSSION TITLE
First user experiences

### DISCUSSION THREAD

**Discussion by Procrat** [enso-org/enso#5955] [Ideas and feedback]
Having had my first experience with the Enso product yesterday, I figured it would be good to share my first experiences in case it can help improve our product. I only accessed online resources (so mostly without asking for help) to adequately capture what other new users would encounter. I believe this is a product that's well on its way to being amazing, so don't take these points as criticism.

|  |  |
| ------- | ---------------------------------------------------------------- |
| 🔵/🟡/🔴 | Low/medium/high importance to me as a user |
| 🐛/❔    | Bug / unexpected behaviour                             |

- Building the IDE
  - 🔵❔[[Tracked](https://github.com/enso-org/enso/issues/6008)] It didn't find the right Node version, even though it was installed, not as the system default but with a Node version manager.
    - After putting a .node-version file in the repo root, that worked fine. That way it works for any Node version manager ([`fnm`](https://github.com/Schniz/fnm), [`nvm`](https://github.com/nvm-sh/nvm), [`nvs`](https://github.com/jasongin/nvs), [`n`](https://github.com/tj/n), [`nave`](https://github.com/isaacs/nave), [`nodeenv`](https://github.com/ekalinin/nodeenv), [`nodist`](https://github.com/nullivex/nodist), ...)
      I can make a PR with it, but the version number would be duplicated in `build-config.yaml`, unless the build script were to look for the version number.
  - 🔵❔[[Tracked](https://github.com/enso-org/enso/issues/5171)] It complained that it couldn't find `conda`. I guessed it tried to install flatbuffers/flatc, which is the only mention of `conda` in `CONTRIBUTING.md`. (I hadn't installed it as it was marked as optional in the README.) However, after I installed flatc directly, it still failed with the same error. After digging through the logs, I saw a warning saying `Failed to find flatc in version == 1.12.0. Found version: 23.3.3`.
    Ilya suggested building with a specific release of the engine, e.g.:
    ```
    ./run ide build --backend-source=release --backend-release=latest
    ```

- Running Enso
  - 🟡🐛 It seems to take up half of CPU time (even before a project is opened).
  - 🟡❔[[Tracked](https://github.com/enso-org/enso/issues/5505)] Starting an example project takes a while without any user feedback (e.g. a spinner).
  - Trying out the Stargazers example
    - 🔵🐛 Why is the preview of the `Data.fetch` node empty? Shouldn't it show the returned JSON?
    - 🟡❔ Why can't I edit nodes? Is this a read-only project? However, I can seem to add new ones? It was only later when I watched the intro video (which I almost never do when trying new software), I realised I had to use Ctrl. Just like double-clicking creates a new node, I would have expected that double-clicking a node enters edit mode.
    - 🔴🐛[Tracked [here](https://github.com/enso-org/enso/issues/5232) and [here](https://github.com/enso-org/enso/issues/5356)] When I drag and drop from the bottom of the last node (to create a new node with that as input), it immediately selects `Table.from_rows`, which after a second shows an error message "Method `from_rows` of Table.value could not be found." in a box that completely covers the input box, so that I'm unable to see what I type.
    - 🔵❔ When a project is opened, I don't understand how I can switch to a different project.
  - 🟡🐛 At some point, Enso kept running in the background after closing it (and keeping a CPU core busy at 100%). It didn't stop when I tried to kill it with the TERM, INT or HUP signals. I had to kill it with the KILL signal. The logs printed:
    ```
    OpenJDK 64-Bit Server VM warning: Exception java.lang.ClassCircularityError occurred dispatching signal SIGTERM to handler- the VM may need to be forcibly terminated
    ```
  - 🟡❔ I was typing a string literal into a node at some point. When pressing Enter, I assumed it would commit the change and go out of edit mode, just like in Excel. However, it instead selected some item from the component browser and replaced what I had typed.
  - 🔵❔[[Tracked](https://github.com/enso-org/enso/issues/5354)] When ctrl-clicking on an existing node, I would expect the component browser to show the documentation of the currently selected method.

- On cloud.enso.org
  - 🟡🐛 At some point there was an error saying it got disconnected (even though I didn't lose my internet connection as far as I know). So I reloaded the page and it said
    ```
    Failed to initialize application: Invalid websocket specification: SyntaxError: Failed to construct 'WebSocket': The URL 'undefinedjson' is invalid..
    ```
    (including those two dots at the end).
    Returning back to cloud.enso.org showed that the workspace had stopped, so that's probably what went wrong.
  - 🔵🐛 At some point, a node showed a warning
    ```
    Error:
    org.enso.interpreter.runtime.control.ThreadInterruptedException
    ```
    I'm not sure what went wrong and I can't reproduce it. Refreshing the page fixed it.
  - 🔴🐛 Without any particular trigger, I completely lost my work at some point! It suddenly reverted my whole solution back to the `0.up_to 100 . to_vector .  map .noise` example. This was a particulary frustrating experience.
  - 🔵🐛 Clicking the close button on the top left doesn't do anything.
  - 🔵🐛 When creating a node, everything seems to freeze for a second. The pointer doesn't even update.

- Using the REPL
  - 🔵🐛 Pressing Enter without any input throws an NPE instead of doing nothing.
  - 🟡❔ It's not obvious to me how I can make modules available. In the IDE, `Table.read` was immediately available. In the REPL, I tried:
    ```
    import Standard.Table
    import Table
    ```
    But neither worked.
  - 🟡❔ The last thing I tried was running Python through the REPL:
    ```
    foreign python f """
        print('test')
    ```
    But that also gave a NPE. I gave up on using the REPL at this point.

- Trying a book club challenge
  - 🔵🐛[Tracked [here](https://github.com/enso-org/enso/issues/5901) and [here](https://github.com/enso-org/enso/issues/5902)] On a `Date.parse` node, I selected "Report Error", after which it completed to `Standard.Base.Error.Problem_Behavior.Report_Error` but then it said
    ```
    Method `Problem_Behavior` of Error could not be found.
    ```
  - 🔵🐛 The preview of a `Day_Of_Week` showed `[Object object]`.
  - 🔵🐛 `Table.from_objects` throws an exception for anything that isn't a JS_Object, Number, Boolean, Text or Nothing, but the docs don't mention that.

- Syntax reference
  - 🟡🐛 There's a broken link on "The installation process can be different depending on your operating system, so please find detailed instructions here."
  - 🔵🐛 There's a heading that says "Argument Blocks Blocks".
  - 🔵❔ I didn't understand this:
      > This operation only "looks through" one level of parentheses. This means that (f _ a) is the same as x -> (f x a), rather than (x -> f x a).
    I thought `f x a` already binds tighter than `->`, so don't those parse the same way?

- Component reference
  - 🔴❔ I'm not sure why I can't click on a lot of intermediate nodes on the tree on the left. For example, I was expecting to be able to click on Base -> Data to read up on `Data.read`, but that didn't work.
  - 🟡❔ I tried looking for `File_Format`, which the documentation of `Data.read` mentions, but I couldn't find it. I also couldn't find `Day_Of_Week`.

**Comment by vitvakatu** [comment]
> It didn't find the right Node version, even though it was installed, not as the system default but with a Node version manager. 

That's great! I really want to avoid manually setting node version via nvm when switching between the branches. @mwu-tow do you think the proposed solution is possible? @Procrat please create an issue for that.

> It complained that it couldn't find conda.

We need to update our docs with clear instructions. Installing the correct `flatc` version is extremely difficult on some platforms. I remember we even had an issue somewhere on Github where we were attaching instructions for patching and installing `flatc`. @Procrat please create an issue for that.

> Starting an example project takes a while without any user feedback (e.g. a spinner).

Tracked in https://github.com/enso-org/enso/issues/5505

> Why is the preview of the Data.fetch node empty? Shouldn't it show the returned JSON?

Indeed. @jdunkerley is it a bug?

> I realised I had to use Ctrl.

Fair point, but the decision to use Ctrl is deliberate. I think no actions are needed for now, although we can revisit it in the future.

> When I drag and drop from the bottom of the last node (to create a new node with that as input), it immediately selects Table.from_rows

I think this should be covered by a bunch of tasks for CB. Namely #5232 and #5356 

> When a project is opened, I don't understand how I can switch to a different project.

`Ctrl-O` (or `Cmd-O`) should work, though it is a temporary solution. Later we will do it using a Dashboard.


> At some point, Enso kept running in the background after closing it 

What OS do you use? It is normal for Mac, but shouldn't happen otherwise.

> I was typing a string literal into a node at some point. When pressing Enter, I assumed it would commit the change and go out of edit mode, just like in Excel.

Again, it's a deliberate decision, but we're aware that it might be suboptimal for some users.

> When ctrl-clicking on an existing node, I would expect the component browser to show the documentation of the currently selected method.

Yeah, I would like it too. I think we will cover it with #5354, we had talked about it with @farmaazon

> On cloud.enso.org

Please create an issue for every point you listed, if you can reproduce it.

**Reply by jdunkerley** [reply]
> > Why is the preview of the Data.fetch node empty? Shouldn't it show the returned JSON?
> 
> Indeed. @jdunkerley is it a bug?

Will check over the example set again this week and see what is not working again!

> > When a project is opened, I don't understand how I can switch to a different project.
> 
> `Ctrl-O` (or `Cmd-O`) should work, though it is a temporary solution. Later we will do it using a Dashboard.

One small improvement for here would be to include the new project in the `Ctrl-O` menu. cc: @farmaazon

**Comment by hubertp** [comment]
> On a Date.parse node, I selected "Report Error", after which it completed to Standard.Base.Error.Problem_Behavior.Report_Error but then it said
Method `Problem_Behavior` of Error could not be found.

Known issues https://github.com/enso-org/enso/issues/5901 and https://github.com/enso-org/enso/issues/5902.

Regarding some other backend issues - I would appreciate if you included full logs with stacktraces. Can't do much with oneliners.
